### PR TITLE
refactor TradingPair

### LINKED ISSUE #1163: Refactor TradingPair
Update the constructors and implement custom Decode to enforce the invariant.

### PR DESCRIPTION
**Pull request by wangjj9219**
close #1163

### CODE REVIEW

File: modules/dex/src/mock.rs [hunk 1]
```diff
@@ -37,9 +37,12 @@ pub const AUSD: CurrencyId = CurrencyId::Token(TokenSymbol::AUSD);
 pub const BTC: CurrencyId = CurrencyId::Token(TokenSymbol::RENBTC);
 pub const DOT: CurrencyId = CurrencyId::Token(TokenSymbol::DOT);
 pub const ACA: CurrencyId = CurrencyId::Token(TokenSymbol::ACA);
-pub const AUSD_BTC_PAIR: TradingPair = TradingPair(AUSD, BTC);
-pub const AUSD_DOT_PAIR: TradingPair = TradingPair(AUSD, DOT);
-pub const DOT_BTC_PAIR: TradingPair = TradingPair(DOT, BTC);
+
+lazy_static::lazy_static! {
```
Comment (xlc) [inline modules/dex/src/mock.rs:41]: you can do 
```
parameter_types! {
  pub static AUSD_BTC_PAIR: TradingPair = TradingPair::from_currency_ids(AUSD, BTC).unwrap();
}
```

File: primitives/src/lib.rs [hunk 2]
```diff
@@ -121,30 +121,53 @@ pub enum DataProviderId {
...
+			.expect("shouldn't be invalid! guaranteed by construction")
+	}
+}
+
+impl Decode for TradingPair {
+	fn decode<I: codec::Input>(input: &mut I) -> sp_std::result::Result<Self, codec::Error> {
+		let first = CurrencyId::decode(input)?;
+		let second = CurrencyId::decode(input)?;
+
+		if (first.is_token_currency_id() || first.is_erc20_currency_id())
```
Comment (xlc) [inline primitives/src/lib.rs:163]: Something like this

```
TradingPair::from_currency_ids(first, second)
       .ok_or(codec::Error::from("invalid currency id"))
```

**Review by xlc — APPROVED** [review]

File: modules/auction-manager/src/mock.rs [hunk 3]
```diff
@@ -155,7 +155,7 @@ parameter_types! {
 	pub const DEXPalletId: PalletId = PalletId(*b"aca/dexm");
 	pub const GetExchangeFee: (u32, u32) = (0, 100);
 	pub const TradingPathLimit: u32 = 3;
-	pub EnabledTradingPairs : Vec<TradingPair> = vec![TradingPair::new(AUSD, BTC)];
+	pub EnabledTradingPairs : Vec<TradingPair> = vec![TradingPair::from_currency_ids(AUSD, BTC).unwrap()];
```
Comment (shaunxw) [inline modules/auction-manager/src/mock.rs:158]:
```suggestion
	pub EnabledTradingPairs: Vec<TradingPair> = vec![TradingPair::from_currency_ids(AUSD, BTC).unwrap()];
```

File: modules/cdp-engine/src/mock.rs [hunk 4]
```diff
@@ -220,7 +220,10 @@ parameter_types! {
 	pub const DEXPalletId: PalletId = PalletId(*b"aca/dexm");
 	pub const GetExchangeFee: (u32, u32) = (0, 100);
 	pub const TradingPathLimit: u32 = 3;
-	pub EnabledTradingPairs : Vec<TradingPair> = vec![TradingPair::new(AUSD, BTC), TradingPair::new(AUSD, DOT)];
+	pub EnabledTradingPairs : Vec<TradingPair> = vec![
```
Comment (shaunxw) [inline modules/cdp-engine/src/mock.rs:223]:
```suggestion
	pub EnabledTradingPairs: Vec<TradingPair> = vec![
```

File: modules/cdp-treasury/src/mock.rs [hunk 5]
```diff
@@ -125,7 +125,11 @@ parameter_types! {
 	pub const GetStableCurrencyId: CurrencyId = AUSD;
 	pub const GetExchangeFee: (u32, u32) = (0, 100);
 	pub const TradingPathLimit: u32 = 3;
-	pub EnabledTradingPairs : Vec<TradingPair> = vec![TradingPair::new(AUSD, BTC), TradingPair::new(AUSD, DOT), TradingPair::new(BTC, DOT)];
+	pub EnabledTradingPairs : Vec<TradingPair> = vec![
```
Comment (shaunxw) [inline modules/cdp-treasury/src/mock.rs:128]:
```suggestion
	pub EnabledTradingPairs: Vec<TradingPair> = vec![
```

**Review by shaunxw — APPROVED** [review]
`.unwrap()` usages in runtime and chain spec could be replaced with `expect("proof of not-panic...")`.

**Pull request merged**